### DISCUSSION TITLE
cmake as a new alternative dbld/rules build tool target

### DISCUSSION THREAD

**Discussion by mitzkia** [syslog-ng/syslog-ng#4208] [Q&A] [answered]
I would like to propose to use `cmake` as an alternative build tool in DBLD/rules targets. I would not like to change the current default `make`.  
It could be used only as an alternative build tool, and this would be my use-case:
`export DEFAULT_IMAGE=devshell; dbld/rules cmake-install ; dbld/rules cmake-light-check`

My original case is the following:
- I have created a new local branch with changes
- I have run successfully the following dbld targets on my laptop `dbld/rules make-install ; dbld/rules make-check ; dbld/rules make-light-check`
- After I have created a Github PR it fails in cmake + Light checks, which I can not run locally in native DBLD way.

I am curious what is your opinion in this topic.

Thanks

**Comment by bazsi** [comment, marked as answer]
I usually go into devshell and invoke stuff from there. 

It would be possible to add top level commands to rules to support this, but you would also have to bootstrap cmake, probably in a subdir of /build

I wouldn't like to ask cake to the normal bootstrap sequence in addition to autogen

**Reply by mitzkia** [reply]
I am currently going with this way. 
But for now I postpone to add cmake as a new alternative dbld/rules build target

**Comment by MrAnno** [comment]
We have been maintaining two different build systems for several years, which is not optimal.

IMHO instead of perpetuating the duality, we should completely move to the new one (which requires additional work towards CMake stability and quality).

**Comment by bazsi** [comment]
there are a lot of things cmake does not yet offer.

IIRC these are missing:
* make dist
* DESTDIR based installation?

Breakage:
* packaging scripts use configure and its arguments (just think of all the
distros that need to change their scripts)

I don't like cmake too much but I can work with it.

On Fri, Nov 11, 2022 at 10:48 PM László Várady ***@***.***>
wrote:

> We have been maintaining two different build systems for several years,
> which is not optimal.
>
> IMHO instead of perpetuating the duality, we should completely move to the
> new one.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/discussions/4208#discussioncomment-4120708>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5SHWTDSVT67DSO6UW3WH25MJANCNFSM6AAAAAAR5LEEIQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Bazsi

**Reply by MrAnno** [reply]
CMake does not lack any features as far as I know, we just did not use them (correctly).
For example, DESTDIR just works as usual: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=paho-mqtt-c#n30
make dist can be implemented with CPack, etc.

I don't have a strong opinion on using CMake or Meson instead of Autotools, I believe all of them can be used in a modern way (they all have a "modern subset"). Even autotools can be clean and modern if it is written carefully.

What I wanted to say is that I don't think it's a good idea to spread the Autotools/CMake duality any further.

**Reply by eli-schwartz** [reply]
Both autotools and Meson offer dist (and more importantly -- dist *check*!) as an out of the box experience. While I suppose it's possible to implement using cpack, that's a bit like saying you can just add a build rule that invokes `tar -cf` in a shell script.

**Comment by mitzkia** [comment]
Thanks for all answers. 
Currently I 'workaround' it with Bazsi's suggestion building inside devshell (manually).
My biggest problem with `cmake` now is that from CMakeList.txt I can not expose an environment variable to the child process (Light).
[Cmake set Docs](https://cmake.org/cmake/help/latest/command/set.html#set-environment-variable)
Like it works smoothly in case of `Makefile.am` with a simple `export [variable]` command.
My goal would be to pass the actual syslog-ng's project source dir to Light (while using (c)make targets).

**Reply by MrAnno** [reply]
It's documented weirdly, but `execute_process()` works with env variables set via `set(ENV)`:

https://cmake.org/pipermail/cmake/2011-August/045665.html
